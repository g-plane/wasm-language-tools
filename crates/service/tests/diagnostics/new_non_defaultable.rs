use super::*;
use insta::assert_json_snapshot;
use wat_service::LanguageService;

#[test]
fn structs() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (type (struct (field)))
  (type $s1 (struct (field i32)))
  (type $s2 (struct (field (ref $s1) i32) (field $f (ref $s1))))
  (func
    struct.new_default 0
    drop
    struct.new_default $s1
    drop
    struct.new_default $s2
    drop))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}

#[test]
fn array() {
    let uri = "untitled:test".to_string();
    let source = "
(module
  (type (array))
  (type $a1 (array (mut i32)))
  (type $a2 (array (ref $a1)))
  (func
    i32.const 0
    array.new_default 0
    drop
    i32.const 0
    array.new_default $a1
    drop
    i32.const 0
    array.new_default $a2
    drop))
";
    let mut service = LanguageService::default();
    service.commit(uri.clone(), source.into());
    calm(&mut service, uri.clone());
    let response = service.pull_diagnostics(create_params(uri));
    assert_json_snapshot!(response);
}
