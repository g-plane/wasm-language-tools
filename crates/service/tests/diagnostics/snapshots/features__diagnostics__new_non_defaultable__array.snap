---
source: crates/service/tests/diagnostics/new_non_defaultable.rs
expression: response
---
{
  "kind": "full",
  "items": [
    {
      "range": {
        "start": {
          "line": 2,
          "character": 14
        },
        "end": {
          "line": 2,
          "character": 15
        }
      },
      "severity": 1,
      "code": "syntax/field-type",
      "source": "wat",
      "message": "syntax error: expected field type"
    },
    {
      "range": {
        "start": {
          "line": 13,
          "character": 22
        },
        "end": {
          "line": 13,
          "character": 25
        }
      },
      "severity": 1,
      "code": "new-non-defaultable",
      "source": "wat",
      "message": "array type `$a2` is not defaultable"
    }
  ]
}
